### ISSUE TITLE
Feature request: a special "<text>" value for the hooks.yaml "files" field

### ISSUE THREAD

**Issue by Lucas-C** [pre-commit/pre-commit#220]
Hi,

I had to implement this kind of "binary-files-filtering-out" today:
https://github.com/Lucas-C/pre-commit-hooks/blob/master/pre_commit_hooks/utils.py#L10

So I made a quick overview of the existing checkers, and I found only 2 that are targeting **any** text file: [end-of-file-fixer](https://github.com/pre-commit/pre-commit-hooks/blob/master/hooks.yaml#L76) and [trailing-whitespaces](https://github.com/pre-commit/pre-commit-hooks/blob/master/hooks.yaml#L106). Both use a long list of known text files suffixes.

If ever you think this use case could come in handy, I'd suggest to provide a magic value like `<any text file>` to the `files` parameter in _hooks.yaml_, so that hooks developers don't have to re-implement this filtering each time.

You can freely re-use the code I mentioned above if it comes handy to you.

Also, sorry for the 3-issues-opened-in-a-row :) I got really excited by this tool and ideas for improvements just flowed in ! But they are really just humble features suggestions.

Cheers

**Comment by asottile** [comment]
+1 this would be great, we already do some file filtering so I imagine we could just add another filtering function to do this. If we didn't support windows I'd probably use something like libmagic, but shouldn't be too hard to find printables / unprintables

**Comment by Lucas-C** [comment]
The solution I used is very portable, take a look :)

We can do even more simple by using the heuristic used by `git`: http://stackoverflow.com/a/7112964/636849
(i.e. just checking for \0 characters)

**Comment by dupuy** [comment]
@asottile wrote:
> If we didn't support windows I'd probably use something like libmagic

In the shell-scripted pre-commit hooks I used before switching to this package, I used `file` magic to select files - this was especially useful for running checks against scripts where there was no file extension, but a simple hashbang check could tell that a particular file was Python, or Perl, or *sh...
I would love it if there were some simple anchored PCRE filter that could select files based on content (of, usually, the first line), so that I could do something like this in `hooks.yaml`:
```
-   id: autopep8-wrapper
    name: autopep8 wrapper
    description: "Runs autopep8 over python source..."
    entry: autopep8-wrapper
    language: python
    files: \.py$
    file-content: #! *?/.*?\bpython\d?\b
    args: [-i]
```
with the hook run against files that match either the filename check or the content check (anchored to start of file content, with `.` not matching newline). This particular regex would match `#! /usr/bin/python` or `#!/usr/bin/env python3.2` or various other kinds of combinations of these sorts of hashbang lines. (The optional space after the bang is allowed - see http://stackoverflow.com/questions/10197690).

**Comment by friederbluemle** [comment]
Any progress on this? If would be very useful to have. Thanks :+1:

**Comment by asottile** [comment]
We've started work on a separate library that will assist in making this possible: https://github.com/chriskuehl/identify

**Comment by asottile** [comment]
@friederbluemle I decided to sit down and complete this today: https://github.com/pre-commit/pre-commit/pull/551

**Comment by friederbluemle** [comment]
Amazing! Thanks @asottile!!

**Comment by Lucas-C** [comment]
Awesome ! Good job :)

**Comment by asottile** [comment]
This is now available in 0.15.0!
